pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -391,11 +391,16 @@ async def _handle_remote_artifacts(self, batch):
                     key = f"{content_artifact.pk}-{d_artifact.remote.pk}"
                     ras_to_create[key] = remote_artifact
 
+        # Make sure we create/update RemoteArtifacts in a stable order, to help
+        # prevent deadlocks in high-concurrency environments. We can rely on the
+        # Artifact sha256 for our ordering.
         if ras_to_create:
-            await sync_to_async(RemoteArtifact.objects.bulk_create)(list(ras_to_create.values()))
+            ras_to_create_ordered = list(ras_to_create.values()).sort(key=lambda x: x.sha256)
```

Comment:
```suggestion
            ras_to_create_ordered = sorted(list(ras_to_create.values()), key=lambda x: x.sha256)
```
Sorry for tricking you into this function that dows not allow chaining commands for no good reason.

Reply:
Heh, no worries - I'd've caught it before pushing, if I hadn't been juggling a few too many balls today :)